src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -0,0 +1,326 @@
+
+        //data is alligned to 512 byte blocks.  a header is 500 bytes, and
+        // the file's data begins exactly after the header
+        // therefor the next header is 500 + o->size rounded up to a mult of 512
+        offset += round_up(o->size + 500);
+        if (newfile_get(&o))         return OS_FS_ERR_DRIVE_NOT_CREATED;
+    }
+
+    //tar ends before two empty records are found
+    return OS_FS_ERROR;
+}
+
+/*
+ * Copies up to req_sz bytes from the tar file into buf
+ */
+int32 tar_read(uint32 offset, char *buf, uint32 req_sz)
+{
+    if (offset + req_sz > size) return OS_FS_ERROR;
+    if (req_sz == 0)            return OS_FS_SUCCESS;
+    uint32 i;
+    for (i = 0 ; i < req_sz ; i++) {
+        buf[i] = start[i + offset];
+    }
+    return OS_FS_SUCCESS;
+}
+
+/*
+ * Copies information from a tar file header to a fsobj
+ */
+uint32 tar_cphdr(uint32 tar_offset, struct fsobj *file)
+{
+    //TODO: store filenames without paths
+    assert(tar_offset < size);
+    assert(file->filedes > 0);
+
+    char *location = start;
+    location += tar_offset;
+    file -> name        = location;
+    if (*(location+strlen(location) - 1) == '/') {
+        file -> type    = FSOBJ_DIR;
+        file -> size    = 0;
+    }
+    else{
+        file -> type    = FSOBJ_FILE;
+        file -> size    = oct_to_dec(location + 124);
+        file -> data    = location + 500;
+    }
+    return OS_FS_SUCCESS;
+}
+
+//this could be more performate bithack
+uint32 round_up(uint32 offset)
+{
+    if (offset % 512) return offset + (512 - (offset % 512));
+    return offset;
+}
+
+/******************************************************************************
+** fsobj Level Methods
+******************************************************************************/
+
+uint32 fsobj_init(struct fsobj *o)
+{
+    assert(o);
+    assert(o->filedes > 0);
+
+    o->name     = "EXAMPLE_NAME";
+	o->type     = FSOBJ_DIR;
+	o->size     = 0;
+    o->refcnt   = 0;
+	o->data     = NULL;
+	o->child    = o->parent = NULL;
+    o->next     = o->prev = NULL;
+    return OS_FS_SUCCESS;
+}
+
+//finds the next open file
+uint32 newfile_get(struct fsobj **o)
+{
+    uint32 count = 0;
+    while (count < MAX_NUM_FILES && files[count].filedes) {
+        count++;
+    }
+    if (count == MAX_NUM_FILES) return OS_FS_ERROR;
+    *o =  &files[count];
+    (*o) -> filedes=count+1;
+    //filedes needs to be unique and nonzero, so filedes is defined as index+1
+    return OS_SUCCESS;
+}
+
+uint32 file_insert(struct fsobj *o, struct FS *filesys)
+{
+    assert(filesys && filesys->root && filesys->root->filedes);
+    if (filesys->root == o) return OS_SUCCESS;
+
+    insert(o, filesys->root);
+
+    return OS_SUCCESS;
+}
+
+//could be written iteratively to use less stackframe if that becomes an issue
+// Root is
+uint32 insert(struct fsobj *o, struct fsobj *root)
+{
+    assert(root && o);
+    assert(strlen(root->name) < strlen(o->name));
```

Comment:
No, it is an internal function.  Checking that an application has passed valid names is done before this.  This assertion is checking the internal consistency of file system code
=